### PR TITLE
add eofmap, handle eof in <-bytes<-unmarshal

### PR DESCRIPTION
**Pull request by AlexKnauth**

### CODE REVIEW

File: io.ss [hunk 1]
```diff
@@ -41,19 +41,24 @@
 (def (bytes<-<-marshal marshal)
   (lambda (x) (call-with-output-u8vector (lambda (port) (marshal x port)))))
 
-;; : ('a <- Bytes) <- ('a <- In)
+;; : ('a <- Bytes) <- ((OrEof 'a) <- In)
```
Comment (fare) [inline io.ss:44]: Why hijack this function with a different type? That seems very wrong.

File: io.ss [hunk 2]
```diff
@@ -41,19 +41,24 @@
...
 (def (marshal<-bytes<- bytes<-)
   (lambda (x port) (write-u8vector (bytes<- x) port)))
 
 ;; : ('a <- In) <- ('a <- Bytes) Nat
 (def (unmarshal<-<-bytes <-bytes n)
-  (lambda (port) (<-bytes (read-bytes* n port))))
+  (lambda (port) (eofmap <-bytes (read-bytes* n port))))
+
+;; eofmap : [A -> B] [OrEof A] -> [OrEof B]
+(def (eofmap f m) (if (eof-object? m) m (f m)))
```
Comment (fare) [inline io.ss:61]: Why do we want that at all??? I don't get it.

**Review by fare — COMMENTED** [review]
I don't get it. This looks like a crock. What we'd rather want are good error messages when (un)marshaling fails somehow.

**Comment by AlexKnauth** [comment]
I was treating `unmarshal` as following the Scheme convention of `read`-like functions. That's returning eof if the input port is empty.

However the ML convention is different... throwing an exception if the input port is empty. Do you want to switch away from the Scheme convention toward the ML convention?

**Comment by AlexKnauth** [comment]
Closing in favor of #25

**Pull request closed without merging**